code/addons/onboarding/src/Onboarding.tsx

```diff
@@ -0,0 +1,276 @@
+const theme = convert();
+
+export type StepKey = (typeof STORYBOOK_ADDON_ONBOARDING_STEPS)[number];
+export type StepDefinition = {
+  key: StepKey;
+  hideNextButton?: boolean;
+  onNextButtonClick?: () => void;
+} & Partial<
+  Pick<
+    // Unfortunately we can't use ts-expect-error here for some reason
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore Ignore circular reference
+    Step,
+    | 'content'
+    | 'disableBeacon'
+    | 'disableOverlay'
+    | 'floaterProps'
+    | 'offset'
+    | 'placement'
+    | 'spotlightClicks'
+    | 'styles'
+    | 'target'
+    | 'title'
+  >
+>;
+
+export default function Onboarding({ api }: { api: API }) {
+  const [enabled, setEnabled] = useState(true);
+  const [showConfetti, setShowConfetti] = useState(false);
+  const [step, setStep] = useState<StepKey>('1:Intro');
+
+  const [primaryControl, setPrimaryControl] = useState<HTMLElement | null>();
+  const [saveFromControls, setSaveFromControls] = useState<HTMLElement | null>();
+  const [createNewStoryForm, setCreateNewStoryForm] = useState<HTMLElement | null>();
+  const [createdStory, setCreatedStory] = useState<{
+    newStoryName: string;
+    sourceFileContent: string;
+    sourceFileName: string;
+  } | null>();
+
+  const selectStory = useCallback(
+    (storyId: string) => {
+      try {
+        const { id, refId } = api.getCurrentStoryData() || {};
+        if (id !== storyId || refId !== undefined) api.selectStory(storyId);
+      } catch (e) {}
+    },
+    [api]
+  );
+
+  const disableOnboarding = useCallback(() => {
+    // remove onboarding query parameter from current url
+    const url = new URL(window.location.href);
+    // @ts-expect-error (not strict)
+    const path = decodeURIComponent(url.searchParams.get('path'));
+    url.search = `?path=${path}&onboarding=false`;
+    history.replaceState({}, '', url.href);
+    api.setQueryParams({ onboarding: 'false' });
+    setEnabled(false);
+  }, [api, setEnabled]);
+
+  const completeOnboarding = useCallback(() => {
+    api.emit(STORYBOOK_ADDON_ONBOARDING_CHANNEL, {
+      step: '6:FinishedOnboarding' satisfies StepKey,
+      type: 'telemetry',
+    });
+    selectStory('configure-your-project--docs');
+    disableOnboarding();
+  }, [api, selectStory, disableOnboarding]);
+
+  useEffect(() => {
+    api.setQueryParams({ onboarding: 'true' });
+    selectStory('example-button--primary');
+    api.togglePanel(true);
+    api.togglePanelPosition('bottom');
+    api.setSelectedPanel('addon-controls');
+  }, [api, selectStory]);
+
+  useEffect(() => {
+    const observer = new MutationObserver(() => {
+      setPrimaryControl(document.getElementById('control-primary'));
+      setSaveFromControls(document.getElementById('save-from-controls'));
+      setCreateNewStoryForm(document.getElementById('create-new-story-form'));
+    });
+
+    observer.observe(document.body, { childList: true, subtree: true });
+    return () => observer.disconnect();
```

Comment:
Does this approach not affect performance? I could imagine this effect being executed dozens of times.

Reply:
Yes, it'll definitely get triggered a bunch when DOM elements are added/inserted. So far I've not found it to be a problem, at least not on my laptop even in low-power mode. The Storybook manager UI doesn't really change that much anyway. Do you think we need to throttle it?